libretro-common/playlists/label_sanitization.c

```diff
@@ -0,0 +1,168 @@
+ *
+ * ---------------------------------------------------------------------------------------
+ * The following license statement only applies to this file (file_path.h).
+ * ---------------------------------------------------------------------------------------
+ *
+ * Permission is hereby granted, free of charge,
+ * to any person obtaining a copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
+ * and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <playlists/label_sanitization.h>
+#include <retro_miscellaneous.h>
+#include <string/stdstring.h>
+#include <string.h>
+
+/*
+ * Does not work with nested blocks.
+ */
+void label_sanitize(char *label, size_t size, bool (*left)(char*), bool (*right)(char*))
+{
+   bool copy = true;
+   int rindex = 0;
+   int lindex = 0;
+   char new_label[PATH_MAX_LENGTH];
+
+   for (; lindex < size && label[lindex] != '\0'; lindex++)
+   {
+      if (copy)
+      {
+         /* check for the start of the range */
+         if ((*left)(&label[lindex]))
+            copy = false;
+
+         if (copy)
+            new_label[rindex++] = label[lindex];
+      }
+      else if ((*right)(&label[lindex]))
+         copy = true;
+   }
+
+   new_label[rindex] = label[lindex];
+
+   strcpy(label, new_label);
+}
+
+bool left_parens(char *left)
+{
+   return left[0] == '(';
+}
+
+bool right_parens(char *right)
+{
+   return right[0] == ')';
+}
+
+bool left_brackets(char *left)
+{
+   return left[0] == '[';
+}
+
+bool right_brackets(char *right)
+{
+   return right[0] == ']';
+}
+
+bool left_parens_or_brackets(char *left)
+{
+   return left[0] == '(' || left[0] == '[';
+}
+
+bool right_parens_or_brackets(char *right)
+{
+   return right[0] == ')' || right[0] == ']';
+}
+
+bool left_parens_or_brackets_excluding_region(char *left)
+{
+   if (left_parens_or_brackets(left))
+   {
+      if (string_is_equal_fast(&left[1], "Australia", 9)
+         || string_is_equal_fast(&left[1], "Europe", 6)
+         || string_is_equal_fast(&left[1], "Japan", 5)
+         || string_is_equal_fast(&left[1], "USA", 3))
+         return false;
```

Comment:
It would be nice if label sanitisation could apply to more ROMs than just no-intro sets. What I mean here is that regions are also commonly written as:

`(EUR)`
`(JAP)`
`(E)`
`(U)`
`(J)`

We're also missing entries for `Germany`, `Korea`, etc. - and the comparison should probably be case insensitive.

As an aside, it would also be nice if 'non-standard' region strings could be converted into no-intro format (e.g. `(U)` -> `(USA)`) - but this is probably beyond the scope of this PR (so I wouldn't worry about it!)

Reply:
It might be hard to account for false positives, e.g. is `(E, A)` a region declaration for `(Europe, Australia)`, or is it part of a title `Cantabile in Two Keys (E, A)(USA)`?

These things keep me up at night.

Reply:
Yes, I appreciate the difficulty here. I think we have to draw the line at multiple entries separated by commas (since this could also be some kind of language list for European titles).

I would suggest only supporting one region 'code' enclosed between parentheses. At the very most, you could perhaps support the standard specific no-entry tags:

`(USA, Europe)`
`(Japan, USA)`

...but anything more sophisticated would require regex matching, which would add impractical performance overheads.

Reply:
Actually, do you know if we have a list of possible regions? Having people open issues to add regions to the matcher might note be the best way to go about things.

Reply:
Well, all the no-intro regions can be found here: https://datomatic.no-intro.org/stuff/The%20Official%20No-Intro%20Convention%20(20071030).pdf (pages 4-5)

...and the GoodTools regions can be found here: http://datacrystal.romhacking.net/wiki/GoodTools#Country_Codes

I guess that covers all the 'official' labels, but I've also seen `(EUR)` and `(JAP)` used in various places.

TBH, that's quite a lot of region codes! If you covered all of them, I don't think any user could complain  :)

Reply:
Don't forget fan translations and hacks which may use all sorts of arbitrary labels and versions. :)

Reply:
Ah, but I think the aim of sanitisation is to trim away stuff like that :)

Most translation tags are contained within square brackets, which get removed anyway when selecting the 'keep region' option. I don't think this PR should try to handle every possible case - that way madness lies!

Reply:
I agree, but it would be good to take account of possible edge cases and not blow up. :)

Reply:
We've reduced our focus on GoodTools in libretro-database. It's riddled with broken hacks and improper names. Just keeping No-Intro convention is probably best for now, and we could consider GoodTools later on.